File: BLSNavigationController/BLSNavigationController.m
//
//  BLSNavigationControllerViewController.m
//  Airpin
//
//  Created by Thomas Carey on 8/2/14.
//  Copyright (c) 2014 Thomas Carey. All rights reserved.
//

#import "BLSNavigationController.h"

@interface BLSNavigationController ()

@property (weak, nonatomic) UIViewController *viewController;
@property (strong, nonatomic) UIProgressView *progressView;

@end

@implementation BLSNavigationController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.progressView = [[UIProgressView alloc] initWithProgressViewStyle:UIProgressViewStyleBar];
    UIProgressView *progress = self.progressView;
    [self.view addSubview:self.progressView];
    UINavigationBar *navBar = self.navigationBar;
    
    NSLayoutConstraint *constraint;
    
    constraint = [NSLayoutConstraint constraintWithItem:progress
                                              attribute:NSLayoutAttributeBottom
                                              relatedBy:NSLayoutRelationEqual
                                                 toItem:navBar
                                              attribute:NSLayoutAttributeBottom
                                             multiplier:1
                                               constant:-0.5];
    [self.view addConstraint:constraint];
    
    constraint = [NSLayoutConstraint constraintWithItem:progress
                                              attribute:NSLayoutAttributeLeft
                                              relatedBy:NSLayoutRelationEqual
                                                 toItem:navBar
                                              attribute:NSLayoutAttributeLeft
                                             multiplier:1
                                               constant:0];
    [self.view addConstraint:constraint];
    
    constraint = [NSLayoutConstraint constraintWithItem:progress
                                              attribute:NSLayoutAttributeRight
                                              relatedBy:NSLayoutRelationEqual
                                                 toItem:navBar
                                              attribute:NSLayoutAttributeRight
                                             multiplier:1
                                               constant:0];
    [self.view addConstraint:constraint];
    
    self.progressView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.progressView setProgress:0.0 animated:NO];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (UIViewController *)viewController
{
    if (!_viewController) {
        _viewController = self.topViewController;
        while ([_viewController class] == [BLSNavigationController class]) {
            _viewController = ((BLSNavigationController *)_viewController).topViewController;
        }
    }
    
    return _viewController;
}

- (BOOL)navigationBar:(UINavigationBar *)navigationBar shouldPopItem:(UINavigationItem *)item
{
    // See http://stackoverflow.com/questions/6413595/uinavigationcontroller-and-uinavigationbardelegate-shouldpopitem-with-monotouc/7453933#7453933
    // for future reference in getting this to work
    
    if ([self.viewController.view respondsToSelector:@selector(canGoBack)] &&
        [self.viewController.view respondsToSelector:@selector(goBack)]) {
        WKWebView *webView = (WKWebView *)self.viewController.view;
        if ([webView canGoBack]) {
            [webView goBack];
            return NO;
        } else {
            [self popViewControllerAnimated:YES];
        }
    } else {
        [self popViewControllerAnimated:YES];
    }
    
    return YES;
}

- (void)updateWithProgress:(float)progress title:(NSString *)title
{
    if (progress == 1.0) {
        [self finishNavigation];
    } else {
        [self updateProgress:progress];
    }
    
    if (title.length) {
        self.viewController.title = title;
    }
}

- (void)updateProgress:(float)progress
{
    if (progress > self.progressView.progress) {
        [self.progressView setProgress:progress animated:YES];
    }
}

- (void)finishNavigation
{
    [UIView animateWithDuration:0.33
                          delay:0
         usingSpringWithDamping:1.0
          initialSpringVelocity:1.0
                        options:0
                     animations:^{
                         [self.progressView setProgress:1.0 animated:YES];
                     }
                     completion:^(BOOL finished) {
                         [UIView animateWithDuration:0.66
                                               delay:0
                              usingSpringWithDamping:1.0
                               initialSpringVelocity:1.0
                                             options:0
                                          animations:^{
                                              self.progressView.alpha = 0.0;
                                          }
                                          completion:^(BOOL finished) {
                                              [self.progressView setProgress:0.0 animated:NO];
                                              self.progressView.alpha = 1.0;
                                          }];
    }];
}

#pragma mark - WKNavigationDelegate

- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(WKNavigation *)navigation
{
    [self updateWithProgress:webView.estimatedProgress title:webView.URL.absoluteString];
}

- (void)webView:(WKWebView *)webView didCommitNavigation:(WKNavigation *)navigation
{
    [self updateWithProgress:webView.estimatedProgress title:webView.title];
}

- (void)webView:(WKWebView *)webView didReceiveServerRedirectForProvisionalNavigation:(WKNavigation *)navigation
{
    [self updateWithProgress:webView.estimatedProgress title:webView.title];
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation
{
    [self updateWithProgress:webView.estimatedProgress title:webView.title];
}

@end

File: BLSNavigationController/BLSNavigationController.swift
//
//  BLSNavigationController.swift
//  BLSNavigationController
//
//  Created by Thomas Carey on 12/1/15.
//  Copyright © 2015 BLS. All rights reserved.
//

import WebKit

class BLSNavigationController: UINavigationController, UINavigationBarDelegate {
  
  let progressView = UIProgressView(progressViewStyle: .Bar)
  
  var webView: WKWebView? {
    didSet {
      if let _ = webView {
        webView?.addObserver(self, forKeyPath: "estimatedProgress", options: .New, context: nil)
        webView?.addObserver(self, forKeyPath: "title", options: .New, context: nil)
      }
    }
    willSet {
      if newValue == nil {
        webView?.removeObserver(self, forKeyPath: "estimatedProgress")
        webView?.removeObserver(self, forKeyPath: "title")
      }
    }
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    configureView()
    configureConstraints()
  }
  
  private func configureView() {
    progressView.translatesAutoresizingMaskIntoConstraints = false
    view.addSubview(progressView)
  }
  
  private func configureConstraints() {
    NSLayoutConstraint(item: progressView, attribute: .Bottom, relatedBy: .Equal, toItem: navigationBar, attribute: .Bottom, multiplier: 1.0, constant: -0.5).active = true
    NSLayoutConstraint(item: progressView, attribute: .Leading, relatedBy: .Equal, toItem: navigationBar, attribute: .Leading, multiplier: 1.0, constant: 0.0).active = true
    NSLayoutConstraint(item: progressView, attribute: .Trailing, relatedBy: .Equal, toItem: navigationBar, attribute: .Trailing, multiplier: 1.0, constant: 0.0).active = true
  }
  
  func navigationBar(navigationBar: UINavigationBar, shouldPopItem item: UINavigationItem) -> Bool {
    // See http://stackoverflow.com/questions/6413595/uinavigationcontroller-and-uinavigationbardelegate-shouldpopitem-with-monotouc/7453933#7453933
    // for future reference in getting this to work  
    
    if let webView = topViewController?.view as? WKWebView {
      if webView.canGoBack {
        webView.goBack()
        return false
      } else {
        self.webView = nil
        popViewControllerAnimated(true)
      }
    } else {
      webView = nil
      popViewControllerAnimated(true)
    }
    
    return true
  }
  
  private func updateProgress(progress: Float) {
    if progress == 1.0 {
      finishNavigation()
    } else {
      progressView.setProgress(progress, animated: true)
    }
  }
  
  private func finishNavigation() {
    UIView.animateWithDuration(0.33, delay: 0.0, usingSpringWithDamping: 1.0, initialSpringVelocity: 1.0, options: [], animations: { () -> Void in
      self.progressView.setProgress(1.0, animated: true)
      self.progressView.alpha = 0.0
      }, completion: { finished in
        self.progressView.setProgress(0.0, animated: false)
        self.progressView.alpha = 1.0
    })
  }
  
  override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
    if keyPath == "estimatedProgress" {
      if let webView = webView {
        updateProgress(Float(webView.estimatedProgress))
      }
    } else if keyPath == "title" {
      let newTitle = change?["new"] as? String
      topViewController?.title = newTitle
    }
  }
}
